skipchain/skipchain.go

```diff
@@ -1360,23 +1369,117 @@ func (s *Service) startBFT(proto string, origRoster, newRoster *onet.Roster, msg
 	}
 }
 
-// PropagateSkipBlock will save a new SkipBlock
-func (s *Service) propagateSkipBlock(msg network.Message) {
-	sbs, ok := msg.(*PropagateSkipBlocks)
+// PropagateGenesisHandler will save a new SkipBlock
+func (s *Service) propagateGenesisHandler(msg network.Message) {
+	pg, ok := msg.(*PropagateGenesis)
 	if !ok {
 		log.Error("Couldn't convert to slice of SkipBlocks")
 		return
 	}
-	for _, sb := range sbs.SkipBlocks {
-		if !s.blockIsFriendly(sb) {
-			log.Lvlf2("%s: block is not friendly: %x", s.ServerIdentity(), sb.Hash)
+
+	if !s.blockIsFriendly(pg.Genesis) {
+		log.Error("Block is not friendly")
+		return
+	}
+
+	id := s.db.Store(pg.Genesis)
+	if id == nil {
+		// the error is logged in the store function
+		return
+	}
+
+	if !pg.Genesis.ParentBlockID.IsNull() {
+		sb := s.db.GetByID(pg.Genesis.ParentBlockID)
+		if sb == nil {
+			log.Lvl3("Unknown parent block. Conode could be out of sync.")
 			return
 		}
+
+		sb.ChildSL = append(sb.ChildSL, pg.Genesis.Hash)
+		s.db.Store(sb)
+	}
+}
+
+// PropagateForwardLinkHandler will update the latest block with
+// the new forward link and the new block when given
+func (s *Service) propagateForwardLinkHandler(msg network.Message) {
+	pfl, ok := msg.(*PropagateForwardLink)
+	if !ok {
+		log.Error("Couldn't convert to a ForwardLink propagation")
+		return
 	}
-	_, err := s.db.StoreBlocks(sbs.SkipBlocks)
+
+	// Get the block to update the list of FLs
+	sb := s.db.GetByID(pfl.ForwardLink.From)
+	if sb == nil {
+		// Here we assume the block must be there because it should
+		// have caught up during the signature request
+		log.Error("Couldn't get the block to attach the forward link")
+		return
+	}
+
+	err := sb.AddForwardLink(pfl.ForwardLink, pfl.Height)
 	if err != nil {
 		log.Error(err)
+		return
 	}
+
+	// Update the forward link list of the block. The link is verified
+	// before the update.
+	id := s.db.Store(sb)
+	if id == nil {
+		// error already logged
+		return
+	}
+
+	// Add the block if available
+	if sb, ok := s.blockBuffer.get(sb.SkipChainID(), pfl.ForwardLink.To); ok {
+		if !s.blockIsFriendly(sb) {
```

Comment:
Is that possible? We have the genesis block, so I think `blockIsFriendly` will return true anyway.